### PR TITLE
consensus/istanbul: add proposal seal in message

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: consensus/istanbul/core/core.go [hunk 1]
```diff
@@ -86,9 +86,20 @@ type core struct {
...
 func (c *core) finalizeMessage(msg *message) ([]byte, error) {
+	var err error
 	// Add sender address
 	msg.Address = c.Address()
 
+	// Add proof of consensus
+	msg.ProposalSeal = []byte{}
+	// A sanity check
+	if c.current.Proposal() != nil {
+		msg.ProposalSeal, err = c.backend.Sign(c.current.Proposal().Hash().Bytes())
```
Comment (bailantaotao) [inline consensus/istanbul/core/core.go]: should add MsgCode

File: consensus/istanbul/core/core.go [hunk 2]
```diff
@@ -256,3 +269,17 @@ func (c *core) newRoundChangeTimer() {
 		c.sendRoundChange()
 	})
 }
+
+// prepareProposalSeal returns a slice from the given hash, number, and msgCode. Replaces msgCode
+// with msgPreprepare if msgCode is lager than uint8.
+func prepareProposalSeal(hash common.Hash, number *big.Int, msgCode uint64) []byte {
+	// a sanity check
+	if msgCode > math.MaxUint8 {
```
Comment (markya0616) [inline consensus/istanbul/core/core.go]: Do we need check this?

Comment (bailantaotao) [inline consensus/istanbul/core/core.go]: 正常來說不會有這個case，但是msgCode如果超過byte長度的話，我後面在做轉換時會panic
```
buf.Write([]byte{byte(msgCode)})
```
不然我就要將uint64轉乘byte array，覺得轉byte array太麻煩，且不需要

File: consensus/istanbul/core/core.go [hunk 3]
```diff
@@ -267,3 +269,16 @@ func (c *core) newRoundChangeTimer() {
 		c.sendRoundChange()
 	})
 }
+
+// prepareProposalSeal returns a slice from the given hash, and msgCode. Replaces msgCode
+// with msgPreprepare if msgCode is lager than uint8.
```
Comment (alanchchen) [inline consensus/istanbul/core/core.go]: larger

Comment (markya0616) [inline consensus/istanbul/core/core.go]: OK

File: consensus/istanbul/core/types.go [hunk 4]
```diff
@@ -94,22 +95,23 @@ type message struct {
 
 // EncodeRLP serializes m into the Ethereum RLP format.
 func (m *message) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, []interface{}{m.Code, m.Msg, m.Address, m.Signature})
+	return rlp.Encode(w, []interface{}{m.Code, m.Msg, m.Address, m.Signature, m.ProposalSeal})
```
Comment (markya0616) [inline consensus/istanbul/core/types.go]: rename to messageSeal

**Pull request merged**